_posts/2020-03-02-tomcat-thread-model.md

```diff
@@ -0,0 +1,96 @@
+
+Opisana sytuacja skłoniła nas do sprawdzenia czy jesteśmy w stanie zapobiec takiemu zachowaniu Tomcata poprzez modyfikację konfiguracji connectorów. Pojawił się między innymi pomysł zmiany implementacji connectora blokującego (bio) na nieblokujący (nio). W systemie, którego dotyczył problem używany był tomcat 7.x, w którym domyślnie używana jest implementacja blokująca.
+
+## Konfiguracja connectorów
+
+Podstawowa konfiguracja Tomcata znajduje się w pliku conf/server.xml i zawiera konfigurację następujących elementów:
+* [Server](http://tomcat.apache.org/tomcat-7.0-doc/config/server.html) - konfiguracja całego kontenera, zawiera konfiguracje poszczególnych serwisów.
+* [Service](http://tomcat.apache.org/tomcat-7.0-doc/config/service.html) - Zawiera konfigurację poszczególnych connectorów łącząc je silnikiem przetwarzania żądań.
+* [Connector](http://tomcat.apache.org/tomcat-7.0-doc/config/http.html) - nasłuchuje na wybranym porcie http i obsługuje połączenia przekazując żądania do silnika zdefiniowanego w elemencie Engine.
+* [Engine](http://tomcat.apache.org/tomcat-7.0-doc/config/engine.html) - przetwarza żądania pochodzące ze wszystkich connectorów zdefiniowanych w elemencie Service.
+
+Z punktu widzenia przetwarzania żądań kluczowe znaczenie ma konfiguracja connectora i parametry:
+* protocol - pozwala określić wybraną implementacje connectora (to tutaj możemy zdecydować czy chcemy użyć connectora blokującego czy nieblokującego).
+* acceptCount - długość kolejki oczekujących połączeń. Kolejka napełnia się jeżeli wszystkie wątki w puli wątków connectora są zajęte. W przypadku osiągnięcia limitu tej kolejki (domyślnie 100) połączenia klientów będą odrzucane.
+* maxConnections - maksymalna liczba połączeń, które mogą być przetwarzane. Po osiągnięciu maksymalnej liczby połączeń, połączenia nadal są przyjmowane i kolejkowane do czasu osiągnięcia limitu wynikającego z parametru acceptCount. Domyślnie maxConnections jest równe maxThreads dla connectora blokującego i 10000 dla nieblokującego.
+* maxThreads - maksymalna liczba wątków obsługujących żądania. Parametr ten oznacza ile żądań może być symultanicznie przetwarzanych przez serwer. Domyślna wartość tego parametru to 200. Biorąc pod uwagę problem, którym zajmujemy się w tym artykule jest to najważniejszy parametr.
+
+## Monitorowanie puli wątków
+
+Aktualna liczba wątków przetwarzających żądania jest obok rozmiaru sterty jednym z najważniejszych parametrów, które powinny być objęte monitoringiem. W tym przypadku monitorujemy atrybut `currentThreadsBusy` mbean'a `Catalina:name="http-bio-8080",type=ThreadPool` (nazwa puli może być inna, np.: http-nio-8080 dla connectora nieblokującego, można to sprawdzić listując wszystkie mbeany lub sprawdzając nazwę wątku w logu). 
```

Comment:
przecinki, które odgradzają to wtrącenie:
Aktualna liczba wątków przetwarzających żądania jest obok rozmiaru sterty jednym
==>
Aktualna liczba wątków przetwarzających żądania jest, obok rozmiaru sterty, jednym

Reply:
Poprawione.